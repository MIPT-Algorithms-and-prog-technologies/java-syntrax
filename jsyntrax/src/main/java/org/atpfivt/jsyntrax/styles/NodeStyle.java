package org.atpfivt.jsyntrax.styles;

import java.awt.Color;
import java.awt.Font;
import java.util.regex.Pattern;

public class NodeStyle {
    public final boolean match(String txt) {
        if (getPattern() == null) {
            return true;
        }
        return getPattern().matcher(txt).matches();
    }

    // gets plain text from regex groups, defined in pattern by user
    // returns text "as is" if no groups found or pattern is not defined
    public final String unwrapTextContent(String txt) {
        if (getPattern() == null) {
            return txt;
        }
        var matcher = getPattern().matcher(txt);
        StringBuilder unwrappedTxt = new StringBuilder();

        if (matcher.groupCount() == 0 || !matcher.find()) {
            return txt;
        }
        for (int i = 1; i <= matcher.groupCount(); i++) {
            unwrappedTxt.append(matcher.group(i));
        }
        return unwrappedTxt.toString();
    }

    private String name = "unknown";
    private String shape = "bubble";
    private Font font = new Font("Sans", Font.BOLD,  14);
    private Color textColor = new Color(0, 0, 0);
    private Color fill = new Color(144, 164, 174);
    private Pattern pattern = null;

    public final String getName() {
        return name;
    }

    public final void setName(String name) {
        this.name = name;
    }

    public final String getShape() {
        return shape;
    }

    public final void setShape(String shape) {
        this.shape = shape;
    }

    public final Font getFont() {
        return font;
    }

    public final void setFont(Font font) {
        this.font = font;
    }

    public final Color getTextColor() {
        return textColor;
    }

    public final void setTextColor(Color textColor) {
        this.textColor = textColor;
    }

    public final Color getFill() {
        return fill;
    }

    public final void setFill(Color fill) {
        this.fill = fill;
    }

    public final Pattern getPattern() {
        return pattern;
    }

    public final void setPattern(Pattern pattern) {
        this.pattern = pattern;
    }
}
