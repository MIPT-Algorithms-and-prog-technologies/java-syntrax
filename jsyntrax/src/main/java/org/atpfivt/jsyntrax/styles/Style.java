package org.atpfivt.jsyntrax.styles;

import java.awt.Color;
import java.awt.Font;
import java.nio.file.Path;
import java.util.List;

public class Style {
    public Style(double scale, boolean transparent) {
        updateByConfig(new StyleConfig());
        this.scale = scale;
        this.transparent = transparent;
    }

    public NodeStyle getNodeStyle(String txt) {
        for (NodeStyle ns : nodeStyles) {
            if (ns.match(txt)) {
                return ns;
            }
        }
        return defNodeStyle;
    }

    public boolean updateByFile(Path style) {
        StyleConfig cfg;
        try {
           cfg = new StyleConfig(style);
        } catch (Exception e) {
            System.out.println("Got Exception: " + e.getMessage());
            return false;
        }
        updateByConfig(cfg);
        return true;
    }

    private void updateByConfig(StyleConfig cfg) {
        line_width = cfg.line_width;
        line_color = cfg.line_color;
        outline_width = cfg.outline_width;
        padding = cfg.padding;
        max_radius = cfg.max_radius;
        h_sep = cfg.h_sep;
        v_sep = cfg.v_sep;
        arrows = cfg.arrows;
        title_pos = cfg.title_pos;
        bullet_fill = cfg.bullet_fill;
        text_color = cfg.text_color;
        shadow = cfg.shadow;
        shadow_fill = cfg.shadow_fill;
        title_font = cfg.title_font;

        nodeStyles = cfg.nodeStyles;
    }

    public int line_width;
    public Color line_color;
    public int outline_width;
    public int padding;
    public int max_radius;
    public int h_sep;
    public int v_sep;
    public boolean arrows;
    public TitlePosition title_pos;
    public Color bullet_fill;
    public Color text_color;
    public boolean shadow;
    public Color shadow_fill;
    public Font title_font;

    private final double scale;
    private final boolean transparent;

    public List<NodeStyle> nodeStyles;

    private static final NodeStyle defNodeStyle = new NodeStyle();

    public double getScale() {
        return scale;
    }

    public boolean isTransparent() {
        return transparent;
    }
}
