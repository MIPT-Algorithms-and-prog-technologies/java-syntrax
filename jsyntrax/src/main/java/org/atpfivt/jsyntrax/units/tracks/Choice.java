package org.atpfivt.jsyntrax.units.tracks;

import org.atpfivt.jsyntrax.units.Unit;
import org.atpfivt.jsyntrax.visitors.Visitor;

import java.util.ArrayList;

public class Choice extends ComplexTrack {
  public Choice(ArrayList<Unit> units) {
    super(units);
  }

  public void accept(Visitor visitor) {
    visitor.visitChoice(this);
  }
}
