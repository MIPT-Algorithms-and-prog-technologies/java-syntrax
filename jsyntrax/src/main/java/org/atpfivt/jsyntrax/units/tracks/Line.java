package org.atpfivt.jsyntrax.units.tracks;

import org.atpfivt.jsyntrax.units.Unit;
import org.atpfivt.jsyntrax.visitors.Visitor;

import java.util.ArrayList;

public class Line extends Track {
  public Line(ArrayList<Unit> units) {
    super(units);
  }

  public Line(Unit unit) {
    super(new ArrayList<>());
    units.add(unit);
  }

  public void accept(Visitor visitor) {
    visitor.visitLine(this);
  }
}
