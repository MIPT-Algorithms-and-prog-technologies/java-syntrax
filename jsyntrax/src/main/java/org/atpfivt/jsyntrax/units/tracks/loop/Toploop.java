package org.atpfivt.jsyntrax.units.tracks.loop;

import org.atpfivt.jsyntrax.exceptions.LoopNotTwoArgsException;
import org.atpfivt.jsyntrax.units.Unit;
import org.atpfivt.jsyntrax.visitors.Visitor;

import java.util.List;

public class Toploop extends Loop {
  public Toploop(List<Unit> units) throws LoopNotTwoArgsException {
    super(units);
  }

  public void accept(Visitor visitor) {
    visitor.visitToploop(this);
  }
}
