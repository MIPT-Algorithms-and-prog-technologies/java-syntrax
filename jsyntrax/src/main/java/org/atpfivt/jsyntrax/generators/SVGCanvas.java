package org.atpfivt.jsyntrax.generators;

import org.atpfivt.jsyntrax.generators.elements.Element;
import org.atpfivt.jsyntrax.styles.NodeStyle;
import org.atpfivt.jsyntrax.styles.Style;
import org.atpfivt.jsyntrax.util.Algorithm;
import org.atpfivt.jsyntrax.util.Pair;

import java.awt.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;

public class SVGCanvas {

    public SVGCanvas(Style style, Map<String, String> urlMap) {
        this.urlMap = urlMap;
        this.style = style;
        this.tagcnt = new HashMap<>();
        this.elements = new ArrayList<>();
    }

    // default prefix = "x", default suffix = ""
    public String new_tag(String prefix, String suffix) {
        String f = prefix + "___" + suffix;
        Integer value = tagcnt.getOrDefault(f, 0);
        tagcnt.put(f, value + 1);
        return prefix + value.toString() + suffix;
    }

    public void addElement(Element e) {
        this.elements.add(e);
    }

    public void addTagByTag(String addTag, String tag) {
        for (Element e : elements) {
            if (e.isTagged(tag)) {
                e.addTag(addTag);
            }
        }
    }

    public void dropTag(final String tag) {
        for (Element e : elements) {
            if (e.isTagged(tag)) {
                e.delTag(tag);
            }
        }
    }

    public void moveByTag(String tag, int dx, int dy) {
        for (Element e : elements) {
            if (e.isTagged(tag)) {
                e.start.f += dx;
                e.start.s += dy;
                e.end.f += dx;
                e.end.s += dy;
            }
        }
    }

    public void scaleByTag(String tag, double scale) {
        for (Element e : elements) {
            if (e.isTagged(tag)) {
                e.scale(scale);
            }
        }
    }

    public Optional<String> getCanvasTag() {
        return elements.stream().findAny().flatMap(Element::getAnyTag);
    }

    public Pair<Pair<Integer, Integer>, Pair<Integer, Integer>> getBoundingBoxByTag(String tag) {
        Pair<Integer, Integer> start = null;
        Pair<Integer, Integer> end = null;
        for (Element e: this.elements)
        {
            if (e.isTagged(tag)) {
                if (start == null) {
                    start = new Pair<>(e.start);
                    end = new Pair<>(e.end);
                }
                start.f = Math.min(start.f, e.start.f);
                start.f = Math.min(start.f, e.end.f);
                start.s = Math.min(start.s, e.start.s);
                start.s = Math.min(start.s, e.end.s);

                end.f = Math.max(end.f, e.start.f);
                end.f = Math.max(end.f, e.end.f);
                end.s = Math.max(end.s, e.start.s);
                end.s = Math.max(end.s, e.end.s);
            }
        }
        return new Pair<>(start, end);
    }

    public String generateSVG() {
        StringBuilder sb = new StringBuilder();
        double scale = style.scale;

        Pair<Pair<Integer, Integer>, Pair<Integer, Integer>> res = getBoundingBoxByTag("all");

        // move to picture to (0, 0)
        moveByTag("all", -res.f.f, -res.f.s);
        scaleByTag("all", scale);
        moveByTag("all", style.padding, style.padding);

        res = getBoundingBoxByTag("all");
        Pair<Integer, Integer> end = res.s;

        int W = end.f + style.padding;
        int H = end.s + style.padding;

        // collect fonts
        HashMap<String, Pair<Font, Color>> fonts = new HashMap<>();
        fonts.put("title_font", new Pair<>(this.style.title_font, this.style.text_color));
        for (NodeStyle ns : this.style.nodeStyles) {
            fonts.put(ns.name + "_font", new Pair<>(ns.font, ns.text_color));
        }

        // header
        sb.append("<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\"?>\n")
                .append("<!-- Created by Jyntrax https://github.com/atp-mipt/jsyntrax -->\n");
        sb.append("<svg xmlns=\"http://www.w3.org/2000/svg\"\n");
        sb.append("xmlns:xlink=\"http://www.w3.org/1999/xlink\"\n");
        sb.append("xml:space=\"preserve\"\n");
        sb.append("width=\"").append(W).append("\" ")
                .append("height=\"").append(H).append("\" ")
                .append("version=\"1.1\">\n");
        // styles
        sb.append("<style type=\"text/css\">\n");
        sb.append("<![CDATA[\n");
        // fonts
        for (Map.Entry<String, Pair<Font, Color>> fontPair : fonts.entrySet()) {
            String fontName = fontPair.getKey();
            String fontFamily = fontPair.getValue().f.getName();
            String fontSize = Integer.toString((int)(fontPair.getValue().f.getSize() * scale));
            String fontWeight = "normal";
            if ((fontPair.getValue().f.getStyle() & Font.BOLD) == Font.BOLD) {
                fontWeight = "bold";
            }
            String fontStyle = "normal";
            if ((fontPair.getValue().f.getStyle() & Font.ITALIC) == Font.ITALIC) {
                fontStyle = "italic";
            }

            String hex = Algorithm.toHex(fontPair.getValue().s);

            sb.append(".").append(fontName).append(" ");
            sb.append("{fill:").append(hex).append("; text-anchor:middle;\n");
            sb.append("font-family:").append(fontFamily).append("; ");
            sb.append("font-size:").append(fontSize).append("pt; ");
            sb.append("font-weight:").append(fontWeight).append("; ");
            sb.append("font-style:").append(fontStyle).append("; ");
            sb.append("}\n");
        }
        // other fonts
        sb.append(".label {fill: #000; text-anchor:middle; font-size:16pt; font-weight:bold; font-family:Sans;}\n");
        sb.append(".link {fill: #0D47A1;}\n");
        sb.append(".link:hover {fill: #0D47A1; text-decoration:underline;}\n");
        sb.append(".link:visited {fill: #4A148C;}\n");
        // close
        sb.append("]]>\n</style>\n");
        // defs
        sb.append("<defs>\n");
        sb.append("<marker id=\"arrow\" markerWidth=\"5\" markerHeight=\"4\" ")
                .append("refX=\"2.5\" refY=\"2\" orient=\"auto\" markerUnits=\"strokeWidth\">\n");
        String hex = Algorithm.toHex(this.style.line_color);
        sb.append("<path d=\"M0,0 L0.5,2 L0,4 L4.5,2 z\" fill=\"").append(hex).append("\" />\n");
        sb.append("</marker>\n</defs>\n");

        // elements
        if (!style.transparent) {
            sb.append("<rect width=\"100%\" height=\"100%\" fill=\"white\"/>\n");
        }
        for (Element e : this.elements) {

            if (style.shadow) {
                e.addShadow(sb, this.style);
            }
            e.toSVG(sb, this.style);
        }
        // end
        sb.append("</svg>\n");
        return sb.toString();
    }

    public final Map<String, String> urlMap;

    private final Style style;

    private final HashMap<String, Integer> tagcnt;

    private final ArrayList<Element> elements;
}
