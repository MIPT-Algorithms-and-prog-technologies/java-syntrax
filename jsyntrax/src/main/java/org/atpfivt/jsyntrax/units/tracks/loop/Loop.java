package org.atpfivt.jsyntrax.units.tracks.loop;

import org.atpfivt.jsyntrax.exceptions.LoopNotTwoArgsException;
import org.atpfivt.jsyntrax.units.Unit;
import org.atpfivt.jsyntrax.units.tracks.ComplexTrack;
import org.atpfivt.jsyntrax.units.tracks.Track;
import org.atpfivt.jsyntrax.visitors.Visitor;

import java.util.List;

public class Loop extends ComplexTrack {
  public Loop(List<Unit> units) throws LoopNotTwoArgsException {
    super(units);

    if (units.size() != 2) {
      throw new LoopNotTwoArgsException();
    }
  }

  public Track getForwardPart() {
    return (Track) getUnits().get(0);
  }

  public Track getBackwardPart() {
    return (Track) getUnits().get(1);
  }

  public boolean isForwardNull() {
    return getForwardPart() == null;
  }

  public boolean isBackwardNull() {
    return getBackwardPart() == null;
  }

  public void accept(Visitor visitor) {
    visitor.visitLoop(this);
  }
}
