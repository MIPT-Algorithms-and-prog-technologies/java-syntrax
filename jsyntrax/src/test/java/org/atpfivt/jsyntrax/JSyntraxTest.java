package org.atpfivt.jsyntrax;

import org.approvaltests.Approvals;
import org.atpfivt.jsyntrax.generators.SVGCanvas;
import org.atpfivt.jsyntrax.generators.SVGCanvasBuilder;
import org.atpfivt.jsyntrax.styles.NodeStyle;
import org.atpfivt.jsyntrax.styles.Style;
import org.atpfivt.jsyntrax.units.tracks.Choice;
import org.atpfivt.jsyntrax.units.tracks.Line;
import org.atpfivt.jsyntrax.units.tracks.loop.Loop;
import org.atpfivt.jsyntrax.units.tracks.loop.Toploop;
import org.atpfivt.jsyntrax.units.tracks.opt.Opt;
import org.atpfivt.jsyntrax.units.tracks.opt.Optx;
import org.atpfivt.jsyntrax.units.tracks.stack.Indentstack;
import org.atpfivt.jsyntrax.units.tracks.stack.Rightstack;
import org.atpfivt.jsyntrax.units.tracks.stack.Stack;
import org.junit.jupiter.api.Test;

import java.awt.*;

import static org.atpfivt.jsyntrax.groovy_parser.SyntraxScript.*;


public class JSyntraxTest {

    private final SVGCanvasBuilder canvasBuilder;
    private final Font testFont;

    public JSyntraxTest() {
        testFont = getTestFont();

        Style s = new Style(1, false);
        s.title_font = testFont;

        for (NodeStyle ns : s.nodeStyles) {
            ns.font = testFont;
        }
        canvasBuilder = new SVGCanvasBuilder().withStyle(s);
    }

    private Font getTestFont() {
        try {
            return Font.createFont(Font.TRUETYPE_FONT,
                    JSyntraxTest.class.getResourceAsStream("PTSans-Regular.ttf"))
                    .deriveFont(12.f);
        } catch (Exception e) {
            throw new IllegalStateException(e);
        }
    }

    @Test
    void lineTest() {
        Line l = line('[', "foo", ',', "/bar", ']');
        SVGCanvas c = canvasBuilder.generateSVG(l);
        String result = c.generateSVG();
        Approvals.verify(result);
    }

    @Test
    void difficultLoopTest() {
        Loop l = loop(line("/forward", "path"), line("backward", "path"));
        SVGCanvas c = canvasBuilder.generateSVG(l);
        String result = c.generateSVG();
        Approvals.verify(result);
    }

    @Test
    void toploopTest() {
        Toploop l = toploop(line('(', "forward", ')'), line(')', "backward", '('));
        SVGCanvas c = canvasBuilder.generateSVG(l);
        String result = c.generateSVG();
        Approvals.verify(result);
    }

    @Test
    void choiceTest() {
        Choice choice = choice('A', 'B', 'C');
        SVGCanvas c = canvasBuilder.generateSVG(choice);
        String result = c.generateSVG();
        Approvals.verify(result);
    }

    @Test
    void optTest() {
        Opt opt = opt('A', 'B', 'C');
        SVGCanvas c = canvasBuilder.generateSVG(opt);
        String result = c.generateSVG();
        Approvals.verify(result);
    }

    @Test
    void optxTest() {
        Optx optx = optx('A', 'B', 'C');
        SVGCanvas c = canvasBuilder.generateSVG(optx);
        String result = c.generateSVG();
        Approvals.verify(result);
    }

    @Test
    void stackTest() {
        Stack s = stack(
                line("top", "line"),
                line("bottom", "line")
        );
        SVGCanvas c = canvasBuilder.generateSVG(s);
        String result = c.generateSVG();
        Approvals.verify(result);
    }


    @Test
    void verticalStackTest() {
        Stack s = stack(
                line('A', 'B'),
                opt("bypass"),
                line("finish")
        );
        SVGCanvas c = canvasBuilder.generateSVG(s);
        String result = c.generateSVG();
        Approvals.verify(result);
    }

    @Test
    void indentStackTest() {
        Indentstack indentStack = indentstack(3,
                line("top", "line"),
                line("bottom", "line")
        );
        SVGCanvas c = canvasBuilder.generateSVG(indentStack);
        String result = c.generateSVG();
        Approvals.verify(result);
    }

    @Test
    void rightStackTest() {
        Rightstack rightStack = rightstack(
                line("top", "line", "with", "more", "code"),
                line("bottom", "line")
        );
        SVGCanvas c = canvasBuilder.generateSVG(rightStack);
        String result = c.generateSVG();
        Approvals.verify(result);
    }
}