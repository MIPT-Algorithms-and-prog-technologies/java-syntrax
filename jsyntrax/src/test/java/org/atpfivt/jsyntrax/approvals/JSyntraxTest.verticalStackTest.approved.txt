<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<!-- Created by Jyntrax https://github.com/atp-mipt/jsyntrax -->
<svg xmlns="http://www.w3.org/2000/svg"
xmlns:xlink="http://www.w3.org/1999/xlink"
xml:space="preserve"
width="141" height="164" version="1.1">
<style type="text/css">
<![CDATA[
.title_font {fill:#000000; text-anchor:middle;
font-family:PT Sans; font-size:12pt; font-weight:normal; font-style:normal; }
.token_font {fill:#000000; text-anchor:middle;
font-family:PT Sans; font-size:12pt; font-weight:normal; font-style:normal; }
.box_font {fill:#000000; text-anchor:middle;
font-family:PT Sans; font-size:12pt; font-weight:normal; font-style:normal; }
.bubble_font {fill:#000000; text-anchor:middle;
font-family:PT Sans; font-size:12pt; font-weight:normal; font-style:normal; }
.hex_font {fill:#000000; text-anchor:middle;
font-family:PT Sans; font-size:12pt; font-weight:normal; font-style:normal; }
.label {fill: #000; text-anchor:middle; font-size:16pt; font-weight:bold; font-family:Sans;}
.link {fill: #0D47A1;}
.link:hover {fill: #0D47A1; text-decoration:underline;}
.link:visited {fill: #4A148C;}
]]>
</style>
<defs>
<marker id="arrow" markerWidth="5" markerHeight="4" refX="2.5" refY="2" orient="auto" markerUnits="strokeWidth">
<path d="M0,0 L0.5,2 L0,4 L4.5,2 z" fill="#000000" />
</marker>
</defs>
<rect width="100%" height="100%" fill="white"/>
<path d="M21,34 A13,13 0 0,1 21,8 H31 A13,13 0 0,1 31,34 z" fill="#000000" fill-opacity="0.4980392156862745" />
<path d="M18,31 A13,13 0 0,1 18,5 H28 A13,13 0 0,1 28,31 z" stroke="#000000" stroke-width="2" fill="#b3e5fc" fill-opacity="1.0" />
<text class="bubble_font" x="23" y="23">A</text>
<path d="M74,34 A13,13 0 0,1 74,8 H84 A13,13 0 0,1 84,34 z" fill="#000000" fill-opacity="0.4980392156862745" />
<path d="M71,31 A13,13 0 0,1 71,5 H81 A13,13 0 0,1 81,31 z" stroke="#000000" stroke-width="2" fill="#b3e5fc" fill-opacity="1.0" />
<text class="bubble_font" x="76" y="23">B</text>
<line x1="40" y1="16" x2="54" y2="16" stroke="#000000" stroke-width="2" marker-end="url(#arrow)" />
<path d="M57,98 A13,13 0 0,1 57,72 H99 A13,13 0 0,1 99,98 z" fill="#000000" fill-opacity="0.4980392156862745" />
<path d="M54,95 A13,13 0 0,1 54,69 H96 A13,13 0 0,1 96,95 z" stroke="#000000" stroke-width="2" fill="#b3e5fc" fill-opacity="1.0" />
<text class="bubble_font" x="75" y="87">bypass</text>
<line x1="94" y1="16" x2="112" y2="16" stroke="#000000" stroke-width="2" />
<path d="M121,25 A9,9 0 0,0 112,16" stroke="#000000" stroke-width="2" fill="none"/>
<line x1="121" y1="25" x2="121" y2="41" stroke="#000000" stroke-width="2" />
<path d="M112,50 A9,9 0 0,0 121,41" stroke="#000000" stroke-width="2" fill="none"/>
<line x1="112" y1="50" x2="71" y2="50" stroke="#000000" stroke-width="2" marker-end="url(#arrow)" />
<line x1="67" y1="50" x2="23" y2="50" stroke="#000000" stroke-width="2" />
<path d="M23,50 A9,9 0 0,0 14,59" stroke="#000000" stroke-width="2" fill="none"/>
<line x1="14" y1="59" x2="14" y2="71" stroke="#000000" stroke-width="2" />
<path d="M14,71 A9,9 0 0,0 23,80" stroke="#000000" stroke-width="2" fill="none"/>
<line x1="23" y1="80" x2="37" y2="80" stroke="#000000" stroke-width="2" marker-end="url(#arrow)" />
<path d="M57,162 A13,13 0 0,1 57,136 H93 A13,13 0 0,1 93,162 z" fill="#000000" fill-opacity="0.4980392156862745" />
<path d="M54,159 A13,13 0 0,1 54,133 H90 A13,13 0 0,1 90,159 z" stroke="#000000" stroke-width="2" fill="#b3e5fc" fill-opacity="1.0" />
<text class="bubble_font" x="72" y="151">finish</text>
<line x1="14" y1="71" x2="14" y2="93" stroke="#000000" stroke-width="2" marker-end="url(#arrow)" />
<line x1="14" y1="97" x2="14" y2="123" stroke="#000000" stroke-width="2" />
<line x1="109" y1="80" x2="127" y2="80" stroke="#000000" stroke-width="2" />
<path d="M136,89 A9,9 0 0,0 127,80" stroke="#000000" stroke-width="2" fill="none"/>
<line x1="136" y1="89" x2="136" y2="105" stroke="#000000" stroke-width="2" />
<path d="M127,114 A9,9 0 0,0 136,105" stroke="#000000" stroke-width="2" fill="none"/>
<line x1="127" y1="114" x2="79" y2="114" stroke="#000000" stroke-width="2" marker-end="url(#arrow)" />
<line x1="75" y1="114" x2="23" y2="114" stroke="#000000" stroke-width="2" />
<path d="M23,114 A9,9 0 0,0 14,123" stroke="#000000" stroke-width="2" fill="none"/>
<line x1="14" y1="123" x2="14" y2="135" stroke="#000000" stroke-width="2" />
<path d="M14,135 A9,9 0 0,0 23,144" stroke="#000000" stroke-width="2" fill="none"/>
<line x1="23" y1="144" x2="37" y2="144" stroke="#000000" stroke-width="2" marker-end="url(#arrow)" />
</svg>
